lib/qa/authorities/linked_data/generic_authority.rb

```diff
@@ -0,0 +1,458 @@
+          url.gsub(pattern, value)
+        end
+
+        def apply_replacements(url, rep_count, rep_config, replacements)
+          return url if rep_count <= 0
+          1.upto(rep_count) do |i|
+            rep = rep_config["replacement_#{i}"]
+            param_name = rep["param"]
+            pattern = rep["pattern"]
+            value = replacements[param_name] || rep["default"]
+            url = url.gsub(pattern, value)
+          end
+          url
+        end
+
+        def get_linked_data(url, language, http_accept = 'application/rdf+xml')
+          uri = URI(url)
+          req = Net::HTTP::Get.new(uri)
+          req['Accept'] = http_accept
+          req['Accept-Langauge'] = language unless language.nil?
+
+          res = fetch(uri, req, 3)
+          #       open 'cached_response', 'w' do |io|
+          #         io.write res.body
+          #       end if res.is_a?(Net::HTTPSuccess)
+          build_graph(res.content_type, res.body)
+        end
+
+        def fetch(uri, req, limit = 10)
+          # You should choose a better exception.
+          raise ArgumentError, 'too many HTTP redirects' if limit.zero?
+          response = Net::HTTP.start(uri.hostname, uri.port) { |http| http.request(req) }
+
+          case response
+          when Net::HTTPSuccess then
+            response
+          when Net::HTTPRedirection then
+            location = response['location']
+            warn "redirected to #{location}"
+            fetch(URI(location), req, limit - 1)
+          when Net::HTTPServerError
+            raise Qa::ServiceUnavailable, "#{uri.hostname} on port #{uri.port} is not responding.  Try again later."
+          when Net::HTTPNotFound
+            raise Qa::TermNotFound, "#{uri} Not Found - Term may not exist at LOD Authority."
+          else
+            response.value
+          end
+        end
+
+        def build_graph(content_type, raw_response)
+          case content_type
+          when 'application/rdf+xml'
+            RDF::Graph.new.from_rdfxml raw_response
+          when 'application/json'
+            RDF::Graph.new.from_jsonld raw_response
+          else
+            raise Qa::UnsupportedFormat, "#{content_type} is unsupported. Supported types include rdf+xml and jsonld format."
+          end
+        end
+
+        def parse_term_authority_response(id, graph, language)
+          graph = filter_language(graph, language) unless language.nil?
+          results = extract_selected_preds_for_term(graph)
+          consolidated_results = consolidate_term_results(results)
+          json_results = convert_term_to_json(consolidated_results)
+          termhash = select_json_result_for_id(json_results, id)
+
+          predicates_hash = predicates_with_subject_uri(graph, termhash[:uri])
+          termhash['predicates'] = predicates_hash unless predicates_hash.length <= 0
+          termhash
+        end
+
+        def filter_language(graph, language)
```

Comment:
Can this be refactored to avoid loops through large graphs?

The `language.nil?` check could probably just be moved outside the `#each` call. This would avoid touching every statement when we already know we won't do anything. The rest seems like it might be possible to implement as a query, which would avoid the need to delete statements.